src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+
+            byte* lowerName;
+            int lowerNameSize;
+            byte* namePtr;
+            int nameSize;
+            DocumentKeyWorker.GetLowerKeySliceAndStorageKey(context, name, out lowerName, out lowerNameSize, out namePtr, out nameSize);
+
+            var modifiedTicks = lastModifiedTicks ?? _documentDatabase.Time.GetUtcNow().Ticks;
+
+            Slice contentTypeSlice;
+            DocumentKeyWorker.GetSliceFromKey(context, contentType, out contentTypeSlice);
+
+            Slice keySlice;
+            using (GetAttachmentKey(context, lowerDocumentId, lowerDocumentIdSize, lowerName, lowerNameSize, out keySlice))
+            {
+                var tbv = new TableValueBuilder
+                {
+                    {keySlice.Content.Ptr, keySlice.Size},
+                    newEtagBigEndian,
+                    {namePtr, nameSize},
+                    {contentTypeSlice.Content.Ptr, contentTypeSlice.Size},
+                    modifiedTicks,
+                };
+
+                Slice streamIdentifierSlice;
+                TableValueReader oldValue;
+                if (table.ReadByKey(keySlice, out oldValue))
+                {
+                    throw new InvalidOperationException("Cannot overwrite exisitng attachment.");
+
+                    /*int size;
+                    var pOldEtag = oldValue.Read(1, out size);
+                    var oldEtag = Bits.SwapBytes(*(long*) pOldEtag);
+
+                    if (expectedEtag != null && oldEtag != expectedEtag)
+                        throw new ConcurrencyException(
+                $"Attachment {key} has etag {oldEtag}, but Put was called with etag {expectedEtag}. Optimistic concurrency violation, transaction will be aborted.")
+            {
+                ActualETag = oldEtag,
+                ExpectedETag = expectedEtag ?? -1
+            };
+
+                    var streamIdentifier = oldValue.Read(streamIdentifierPosition, out size);
+                    tbv.Add(streamIdentifier, size);
+                    streamIdentifierScope = Slice.External(context.Allocator, streamIdentifier, size, out streamIdentifierSlice);
+
+                    table.Update(oldValue.Id, tbv);*/
+                }
+                else
+                {
+                    if (expectedEtag.HasValue && expectedEtag.Value != 0)
+                    {
+                        ThrowConcurrentExceptionOnMissingAttacment(documentId, name, expectedEtag.Value);
+                    }
+
+                    tbv.Add(newEtag); // streamIdentifier
+                    tbv.SliceFromLocation(context.Allocator, (int) AttachmentsTable.StreamIdentifier, out streamIdentifierSlice);
+
+                    table.Insert(tbv);
+                }
+
+                // Insert the stream
+                var tree = context.Transaction.InnerTransaction.CreateTree(AttachmentsSlice);
```

Comment:
Might be better to drop the storage id notion. The issue is that we are going to have to do accounting, and we want to be able to show the stream name to the user (if an attachment is very large, we want to be able to tell what it is).
Let us use the attachment key as the stream key as well, so we'll be able to handle it.